server/storage/rethinkdb.go

```diff
@@ -294,7 +329,110 @@ func (rdb RethinkDB) CheckHealth() error {
+	ch := Change{
+		CreatedAt: now,
+		GUN:       gun,
+		Version:   version,
+		SHA256:    sha256,
+		Category:  category,
+	}
+	_, err := gorethink.DB(rdb.dbName).Table(ch.TableName()).Insert(
+		ch,
+		gorethink.InsertOpts{
+			Conflict: "error", // default but explicit for clarity of intent
+		},
+	).RunWrite(rdb.sess)
+	return err
+}
+
+// GetChanges returns up to pageSize changes starting from changeID. It uses the
+// blackout to account for RethinkDB's eventual consistency model
 func (rdb RethinkDB) GetChanges(changeID string, pageSize int, filterName string) ([]Change, error) {
-	return nil, errors.New("Not Implemented")
+	var (
+		lower, upper, bound []interface{}
+		idx                 = "rdb_created_at_id"
+		max                 = []interface{}{gorethink.Now().Sub(blackoutTime), gorethink.MaxVal}
+		min                 = []interface{}{gorethink.MinVal, gorethink.MinVal}
+		order               gorethink.OrderByOpts
+		reversed            bool
+	)
+	if filterName != "" {
+		idx = "rdb_gun_created_at_id"
+		max = append([]interface{}{filterName}, max...)
+		min = append([]interface{}{filterName}, min...)
+	}
+
+	switch changeID {
+	case "0", "-1":
+		lower = min
+		upper = max
+	default:
+		bound, idx = rdb.bound(changeID, filterName)
+		if pageSize < 0 {
+			lower = min
+			upper = bound
+		} else {
+			lower = bound
+			upper = max
+		}
+	}
+
+	if changeID == "-1" || pageSize < 0 {
+		reversed = true
+		order = gorethink.OrderByOpts{Index: gorethink.Desc(idx)}
+	} else {
+		order = gorethink.OrderByOpts{Index: gorethink.Asc(idx)}
+	}
+
+	if pageSize < 0 {
+		pageSize = pageSize * -1
+	}
+
+	changes := make([]Change, 0, pageSize)
+
+	// Between returns a slice of results from the rethinkdb table.
+	// The results are ordered using BetweenOpts.Index, which will
+	// default to the index of the immediately preceding OrderBy.
+	// The lower and upper are the start and end points for the slice
+	// and the Left/RightBound values determine whether the lower and
+	// upper values are included in the result per normal set semantics
+	// of "open" and "closed"
+	res, err := gorethink.DB(rdb.dbName).
+		Table(Change{}.TableName()).
+		OrderBy(order).
+		Between(
+			lower,
+			upper,
+			gorethink.BetweenOpts{
+				LeftBound:  "open",
+				RightBound: "open",
+			},
+		).Limit(pageSize).Run(rdb.sess)
+	if err != nil {
+		return nil, err
+	}
+	defer res.Close()
+
+	defer func() {
+		if reversed {
+			// results are currently newest to oldest, should be oldest to newest
+			for i, j := 0, len(changes)-1; i < j; i, j = i+1, j-1 {
+				changes[i], changes[j] = changes[j], changes[i]
+			}
+		}
+	}()
+
+	return changes, res.All(&changes)
+}
+
+// bound creates the correct boundary based in the index that should be used for
+// querying the changefeed.
+func (rdb RethinkDB) bound(changeID, filterName string) ([]interface{}, string) {
+	createdAtTerm := gorethink.DB(rdb.dbName).Table(Change{}.TableName()).Get(changeID).Field("created_at")
```

Comment:
What happens if the change ID doesn't exist in the table?  (and can we have a test for that case?)

Reply:
Good question. Adding the test case. I think this is where the behaviours may unfortunately diverge between gorm and rethinkdb. RethinkDB will likely error due to the `bound` function looking up an explicit change ID, while MySQL will happily just use the ID as a bound.

Reply:
Ok, test pushed. As far as the API spec, it probably makes sense for us to update it to specify the behavior is "undefined" (gives us a lot of wiggle room) in the cases where a changeID other than `0`, `-1`, or a known valid ID are provided.

Reply:
I added 2 variations of the test to simulate multiple expected behaviours https://github.com/docker/notary/pull/1214/commits/51bcded0bfd67a43bc67c78d78759ebbfb234c0d#diff-523c4d973493ddd421ebe1b81b5ec81eR432

Reply:
Thanks for the update!  Wondering if we should catch these specifically (where the ID can't be coerced to an int in SQL, or that the ID doesn't exist in rethink, and return an `InvalidChangeID` type error?  Otherwise I think in the handler this gets translated into a 500, whereas it should probably be a 400?  Otherwise we might see some spurious reports/pages/etc.